### PR TITLE
Fixed #24638 -- Added QuerySet.comment() to inject SQL comments.

### PR DESCRIPTION
**Pull request by dcsid**
#### Trac ticket number

ticket-24638

#### Branch description

Adds `QuerySet.comment(message)` for placing SQL block comments in generated `SELECT` and `UPDATE` statements. Comments are preserved across queryset cloning/chaining, appear in the expected position for combined queries, and reject SQL comment delimiters to prevent breaking out of the generated comment block.

#### AI Assistance Disclosure (REQUIRED)

- [ ] **No AI tools were used** in preparing this PR.
- [x] **If AI tools were used**, I have disclosed which ones, and fully reviewed and verified their output.

Used Claude Opus 4.7 to help draft the implementation and tests. Used Codex for a follow-up review of the patch, CI status, and Trac/PR metadata. I then manually reviewed the final diff, tests, documentation, and PR description.

#### Checklist

- [x] This PR follows the [contribution guidelines](https://docs.djangoproject.com/en/stable/internals/contributing/writing-code/submitting-patches/).
- [x] This PR **does not** disclose a security vulnerability (see [vulnerability reporting](https://docs.djangoproject.com/en/stable/internals/security/)).
- [x] This PR targets the `main` branch. 
- [x] The commit message is written in past tense, mentions the ticket number, and ends with a period (see [guidelines](https://docs.djangoproject.com/en/dev/internals/contributing/committing-code/#committing-guidelines)).
- [x] I have not requested, and will not request, an automated AI review for this PR. 
- [x] I have checked the "Has patch" ticket flag in the Trac system.
- [x] I have added or updated relevant tests.
- [x] I have added or updated relevant docs, including release notes if applicable.
- [ ] I have attached screenshots in both light and dark modes for any UI changes.

### CODE REVIEW

**Comment by felixxm** [comment]
@dcsid Please don't add LLM generated descriptions with unnecessary details, and don't ping folks directly. Also, I see no reason to mention tools like OpenTelemetry, sqlcommenter, and others in release notes or docstrings.

File: django/db/models/query.py [hunk 1]
```diff
@@ -1972,6 +1975,19 @@ def fetch_mode(self, fetch_mode):
...
 
+    def comment(self, message):
+        """Add an SQL comment to the query."""
+        if not isinstance(message, str):
+            raise TypeError("QuerySet.comment() argument must be a string.")
+        if "/*" in message or "*/" in message:
+            raise ValueError(
+                "QuerySet.comment() cannot include '/*' or '*/'; strip or "
+                "escape these delimiters before calling comment()."
+            )
```
Comment (jacobtylerwalls) [inline django/db/models/query.py:1997]: I'm concerned this is too weak -- we've seen how some C adapters like psycopg(2) aren't resilient against null bytes. Passing in `"\x00"` here, I can cause:

```
ProgrammingError: unterminated /* comment at or near "/* "
```

Comment (dcsid) [reply django/db/models/query.py:1997]: Thanks for catching this. I added validation to reject null bytes in `QuerySet.comment()`, added a regression test for `"\x00"`, and updated the docs to mention this `ValueError` case.

Comment (adamchainz) [reply django/db/models/query.py:1997]: Let's use an "allow list" instead, perhaps a basic regex that allows letters, numbers, spaces, and limited punctuation: `r'^[\w ._-]+$'` ?

I'm just thinking we've played "whack-a-mole" in the security team trying to filter out characters allowed in certain places in the ORM:

* [CVE-2026-1207, CVE-2026-1287, and CVE-2026-1312](https://www.djangoproject.com/weblog/2026/feb/03/security-releases/)
* [CVE-2025-13372](https://www.djangoproject.com/weblog/2025/dec/02/security-releases/)

…and so on.

I'd like to avoid this with the new feature!

File: django/db/models/query.py [hunk 2]
```diff
@@ -1972,6 +1975,19 @@ def fetch_mode(self, fetch_mode):
         clone._fetch_mode = fetch_mode
         return clone
 
+    def comment(self, message):
+        """Add an SQL comment to the query."""
+        if not isinstance(message, str):
+            raise TypeError("QuerySet.comment() argument must be a string.")
+        if "/*" in message or "*/" in message or "\x00" in message:
```
Comment (RealOrangeOne) [inline django/db/models/query.py]: Thought: There are many other characters restricted from `FORBIDDEN_ALIAS_PATTERN`. Obviously some are still valid in comments, but I suspect blocking all control characters rather than just null is probably a good idea.

File: django/db/models/query.py (hunk 1 above)
Comment (dcsid) [reply django/db/models/query.py:1997]: Updated this to use an allowlist for `QuerySet.comment()` messages, limited to word characters, spaces, periods, underscores, and hyphens. I also updated the docs and tests to reject SQL comment delimiters, null bytes, all C0/C1 control characters, and punctuation outside the allowed set.

Does this match the allowlist approach you had in mind, or would you prefer this to be stricter, e.g. ASCII-only?

File: django/db/models/query.py (hunk 2 above)
Comment (dcsid) [reply django/db/models/query.py]: Updated this so validation now uses an allowlist rather than only rejecting specific characters. This rejects all control characters, including null bytes, newlines, tabs, and C1 controls.

Does that address everything, or are there other character classes you think should be rejected explicitly?

File: django/db/models/query.py [hunk 3]
```diff
@@ -47,6 +48,7 @@
 REPR_OUTPUT_SIZE = 20
 
 DEFAULT_FETCH_MODE = FETCH_ONE
+SQL_COMMENT_RE = _lazy_re_compile(r"^[\w ._-]+$")
```
Comment (adamchainz) [inline django/db/models/query.py]: This can be lowercase :)
```suggestion
sql_comment_re = _lazy_re_compile(r"^[\w ._-]+$")
```

File: django/db/models/query.py (hunk 1 above)
Comment (adamchainz) [reply django/db/models/query.py:1997]: Yeah, looks good to me. Hopefully it's not too restrictive for others, though we can always expand the set of allowed characters quite easily.

**Comment by jacobtylerwalls** [comment]
Thanks for the updates. After speaking with the Security Team last week, could I ask for a summary of how this is implemented in peer frameworks/ORMs? If the prior art all points in one direction, this becomes easier to review. If it doesn't, we might want to gather more community input.

**Comment by dcsid** [comment]
Thanks Jacob, that makes sense. I took a look at a few other ORMs/frameworks to see how they handle this:

| Framework / ORM | How they handle SQL comments/query tags |
| --- | --- |
| Rails Active Record | `Relation#annotate()` adds block comments like `/* comment */` to generated SQL. Rails also has Query Logs for SQLCommenter-style observability tags. |
| EF Core | `TagWith()` adds one or more line comments before the generated SQL. Multiple tags are cumulative. |
| SQLAlchemy | `prefix_with()` and `suffix_with()` let users add text around generated SQL statements, including comments or dialect-specific hints. |
| Hibernate | `Query#setComment()` stores a per-query comment that can be included in generated SQL when SQL comments are enabled. Query hints are handled separately. |
| Prisma | Supports SQLCommenter metadata for observability/tracing, configured at the client/plugin level rather than as a per-query ORM method. |

My takeaway is that there is meaningful prior art for attaching comments or tags to ORM-generated SQL for observability/debugging. The exact placement differs by project. Rails appends block comments, EF Core prepends line comments, SQLAlchemy exposes prefix/suffix hooks, and Prisma handles it more globally through SQLCommenter metadata.

This PR seems more conservative than some of those approaches because it restricts the allowed comment text and avoids optimizer-hint syntax or arbitrary SQL fragments.

Sources:
- Rails `annotate()`: https://api.rubyonrails.org/classes/ActiveRecord/QueryMethods.html#method-i-annotate
- Rails Query Logs: https://api.rubyonrails.org/v7.1.0/classes/ActiveRecord/QueryLogs.html
- EF Core `TagWith()`: https://learn.microsoft.com/ef/core/querying/tags
- SQLAlchemy `prefix_with()` / `suffix_with()`: https://docs.sqlalchemy.org/en/20/core/selectable.html
- Hibernate `setComment()`: https://docs.hibernate.org/orm/6.1/javadocs/org/hibernate/query/Query.html
- Prisma SQL comments: https://www.prisma.io/docs/orm/prisma-client/observability-and-logging/sql-comments

**Comment by dcsid** [comment]
Following up on the prior art. Rails and SQLAlchemy accept arbitrary text, EF Core and Hibernate are similarly permissive, and only Prisma constrains the format. This PR is the most restrictive of the group. The allowlist (`^[\w ._-]+$`) avoids the filter-bypass surface flagged in the recent CVEs, since only a small safe set is accepted in the first place.

Happy to make further changes if needed.

**Comment by dcsid** [comment]
Pushed some added test coverage for the `DELETE`, `INSERT`, and combinator paths.
